### PR TITLE
Set up clippy in CI

### PR DESCRIPTION
**Pull request by djc**
This also helped detect remaining naming issues.

### CODE REVIEW

File: rustls/src/keylog.rs [hunk 1]
```diff
@@ -160,6 +159,13 @@ impl KeyLog for KeyLogFile {
     }
 }
 
+impl Default for KeyLogFile {
```
Comment (briansmith) [inline rustls/src/keylog.rs:162]: This is worse. Let's just disable this lint.

File: rustls/src/msgs/codec.rs [hunk 2]
```diff
@@ -120,6 +120,7 @@ impl u24 {
     }
 }
 
+#[allow(clippy::from_over_into)]
```
Comment (briansmith) [inline rustls/src/msgs/codec.rs:123]: It seems like we should follow this PR up with a new one that does what the lint suggests. I think these are only written the way they are because of the old coherence rules.

File: rustls/src/server/handy.rs [hunk 3]
```diff
@@ -135,16 +134,15 @@ impl server::ResolvesServerCert for AlwaysResolvesChain {
 
 /// Something that resolves do different cert chains/keys based
 /// on client-supplied server name (via SNI).
+#[derive(Default)]
```
Comment (briansmith) [inline rustls/src/server/handy.rs:137]: Again, we should undo this and disable the lint.

File: rustls/src/sign.rs [hunk 4]
```diff
@@ -397,3 +399,7 @@ pub fn supported_sign_tls13() -> &'static [SignatureScheme] {
         SignatureScheme::ED25519,
     ]
 }
+
+/// Errors while signing
+#[derive(Debug)]
+pub struct SignError;
```
Comment (briansmith) [inline rustls/src/sign.rs:405]: Is this supposed to be publicly-constructible? I would expect it to be `pub struct SignError(());` or `pub struct SignError(pub(crate) ());`.

**Review by briansmith — CHANGES_REQUESTED** [review]
I'm not convinced that the `SignError` stuff is the best API design but it's no worse than `()`.

File: rustls/src/msgs/codec.rs (hunk 2 above)
Comment (djc) [reply rustls/src/msgs/codec.rs:123]: In my opinion, conceptually it does not make sense to have a conversion from `usize` -> `u24`.

File: rustls/src/sign.rs (hunk 4 above)
Comment (djc) [reply rustls/src/sign.rs]: Fixed.

File: rustls/src/keylog.rs (hunk 1 above)
Comment (djc) [reply rustls/src/keylog.rs]: Fixed.

File: rustls/src/client/hs.rs [hunk 5]
```diff
@@ -226,7 +224,7 @@ fn emit_client_hello_for_retry(
     extra_exts: Vec<ClientExtension>,
     may_send_sct_list: bool,
     suite: Option<&'static SupportedCipherSuite>,
-) -> NextStateOrError {
+) -> Box<dyn State + Send + Sync> {
```
[the remark this answers is not in the document]
Comment (djc) [reply rustls/src/client/hs.rs]: Changed it to `NextState` for consistency. Honestly, while I maybe see the point for the unwieldy `Result<Box<..>, Error>`, I sometimes find the aliases to be unhelpful in that they obscure what's actually there.

**Comment by ctz** [comment]
Merged as d03bf27, thanks!

File: rustls/src/msgs/codec.rs (hunk 2 above)
Comment (briansmith) [reply rustls/src/msgs/codec.rs:123]: I think the `from_over_into` lint is telling you to `impl From<u24> for usize` instead of of `impl Into<usize> for u24`; i.e. the same conversion you're currently doing, but in terms of `From`.

Comment (djc) [reply rustls/src/msgs/codec.rs:123]: Sure, but the reason given is that `From` impls have a blanket impl for the converse `Into` impl, which in this case seems undesirable to me.

Comment (briansmith) [reply rustls/src/msgs/codec.rs:123]: > Sure, but the reason given is that From impls have a blanket impl for the converse Into impl, which in this case seems undesirable to me.

No, it won't automatically generate the "converse." It will instead generate this exact `impl Into<usize> for u24`.

Basically, it's saying "Hey, it looks like you are implementing `Into<usize> for u24`  because the old coherence rules won't let you write `From<u24> for usize`, but the new coherence rules do let you do that, and that's more idiomatic."

Basically we should almost never implement `Into` directly in any case where it's possible to implement the equivalent `From` instead.

Comment (djc) [reply rustls/src/msgs/codec.rs:123]: Hmm, you're right, of course. Implemented this in #631.
